Week2/homework/maartjes-work.js

```diff
@@ -45,21 +45,32 @@ const tuesday = [
 const maartjesTasks = monday.concat(tuesday);
 const maartjesHourlyRate = 20;
 
-function computeEarnings(tasks, hourlyRate) {
-  // Replace this comment and the next line with your code
-  console.log(tasks, hourlyRate);
-}
-
-// eslint-disable-next-line no-unused-vars
-const earnings = computeEarnings(maartjesTasks, maartjesHourlyRate);
-
+// Start: Adham's code
+const maartjesDurationsHours = []; /* Collect our data in one array */
```

Comment:
Hi Adham, after going through your discussion with @remarcmij about tests for this homework and reviewing your `maartjes-work.js`, I think I know what's happening and what possibly was the cause of all the pain you had to go through last week.

Firstly, the reason why your homework didn't pass the tests was because you removed the original function `computeEarnings` in which you were supposed to actually add your code to. The unit tests expect `computeEarnings` to be a valid function in the JS file and for it to return a numerical value. So when the test saw you removed the function entirely, that upset it. I think it was its way of revenge from you. 🤷‍♂ 

Secondly, the reason why we have it like this in HYF's homework is because it closely follows to what developers actually do in real life. This is also known as **Test-Driven Development**. That means, before a developer starts writing code to solve a problem, he/she is required to first write unit tests for the code which will be eventually written.

So if you had put all your logic inside the `computeEarnings` function and simply returned `maartjesEarnings` from it, you wouldn't have had that problem. I am not sure why you decided to remove that function though. 🙂

Reply:
Thanks Yash,
I believe that  the target of unit tests is to test the code for errors and helps the developers to notice problems if they got  unexpected results such as negative value or too big value. But it is not to test if the writer has followed exactly the steps that some one else wants . If that one knows the steps and the result then no need to let me think about solving it.  In that case just give me a list with all the variables, functions and methods that I need to use to follow the wanted steps, and also make the steps very clear , so no need for me to think but just converting the steps to code. (I mean: something like user instruction in user manual, then I just need to press the correct buttons
I like the question very much it is not like all the others nonsense questions.  But the way how it is corrected makes it much worse than others questions,(I mean the way how the test code it written).
I hoop from HYF in the future : if they want me to use a specific method then do not give such a kind real situation practice, but just  an input like aan object (X) and then ask to use Map() and tell me that the output must be Y. so if I am not using map or not getting that output then sure I have misunderstanding about Map(). And no need for such a test file.
I deleted the Function because the test did not accept it . I tried to give the test what it wants, but always the test was returning to  me with nonsense  output ( the x function is not a function) something like that. 
so I decided to delete everything and keep my won code. 
I do not know when I will go back to that question, I have to do that but it will not be soon. 
Best regards.

Reply:
Hi Adham,

I agree that unit tests help developers in catching unexpected results before the code is submitted/shared with others. However, I think you are mistaken in saying that it forces you to follow exactly the steps that someone else wants. Firstly, in real life scenario, it is not someone else who writes unit tests for you. It is the individual developer's responsibility to write tests for their own code. In HackYourFuture, the mentors make it easier for students by writing unit tests for them.

By definition, unit tests are meant to test small units of an application which in a real world can be considered as a function. For example, if I have a function like:

```
const sum = (a,b) => a + b;
```

One will write unit tests for this function such as:

```
describe('testing sum function', () => {
  it('should return 4', () => {
    expect(sum(1,3)).toEqual(4);
  });

  it('should return 10', () => {
    expect(sum(5,5)).toEqual(10);
  });
});
```

No way do these unit tests prevent a developer for implementing their own sum function. All it tests is if I provide my sum function some arguments, do they return a value which is expected or something completely unexpected? I can easily have written the function as:

```
const sum = (num1,num2) => {
  const total = parseInt(num1, 10) + parseInt(num2, 10);
  return total;
};
```

Coming back to the homework, the function being unit tested is `computeEarnings` function.  Test written for this function are in such a way that it expects result to be `373.33` if provided it with `maartjesTasks` and `maartjesHourlyRate`. Again, no way does this test forces you to write code in a specific manner. It will be happy as long as your code returns 373.33 when provided with `maartjesTasks` and `maartjesHourlyRate`.

If I were to copy your code and paste it inside the `computeEarnings` function, it would work just fine and that is what the test expected from you. 🙂 

```
function computeEarnings(maartjesTasks, maartjesHourlyRate) {
  const maartjesDurationsHours = []; /* Collect our data in one array */
  maartjesTasks.forEach(element => {
    const durationMinutes = element.duration;
    const durationHours = durationMinutes / 60;
    if (durationHours >= 2) {
      maartjesDurationsHours.push(durationHours);
    }
  });
  // console.log(maartjesDurationsHours); /* Array of durations in hours, just if the duration is more than 2 hours */
  const reducingProcess = (totalToLeft, right) => totalToLeft + right;
  const maartjesTotalDuration = maartjesDurationsHours.reduce(reducingProcess);
  // console.log(maartjesTotalDuration); /* One Number: the total of durations in hours */
  const maartjesEarnings = maartjesTotalDuration * maartjesHourlyRate;

  return maartjesEarnings;
}

const earnings = computeEarnings(maartjesTasks, maartjesHourlyRate);
```

I can also rewrite the function like below and tests would still pass

```
function computeEarnings(tasks, hourlyRate) {
  const totalEarnings = tasks
    .map(task => ({ name: task.name, duration: task.duration / 60 })) // transforms minutes to hours
    .filter(time => time.duration >= 2) // filter array task duration equal and greater than 2
    .reduce((acc, task) => acc + task.duration * hourlyRate, 0); // add duration * hourly rate values to the acc

  return totalEarnings;
}
```

As you can see, one can write code in any way one wants and the unit test do not restrict me with the approach to be taken.

As far as revisiting homework is concerned, I would suggest you to speak with Wouter, Fede or Noer about it as they would be able to consult you better. But for now, I won't be able to approve or close this PR. See you tomorrow. 🙂